### ISSUE TITLE
invalid application of 'sizeof' to incomplete type 'arm_compute::NEDepthConcatenateKernel'

### ISSUE THREAD

**Issue by SunAriesCN** [ARM-software/ComputeLibrary#220]
When I was building the squeezenet with ACL, I just got a problem as below:
```
In file included from /home/user/my-toolchains/aarch64-linux-android-4.9/include/c++/4.9.x/memory:81:0,
                 from ./arm_compute/runtime/NEON/INESimpleFunction.h:31,
                 from ./arm_compute/runtime/NEON/functions/NEAbsoluteDifference.h:27,
                 from ./arm_compute/runtime/NEON/NEFunctions.h:28,
                 from neon_squeezenet/neon_squeezenet.cpp:1:
/home/user/my-toolchains/aarch64-linux-android-4.9/include/c++/4.9.x/bits/unique_ptr.h: In instantiation of 'void std::default_delete<_Tp []>::operator()(_Tp*) const [with _Tp = arm_compute::NEDepthConcatenateKernel]':
/home/user/my-toolchains/aarch64-linux-android-4.9/include/c++/4.9.x/bits/unique_ptr.h:484:16:   required from 'std::unique_ptr<_Tp [], _Dp>::~unique_ptr() [with _Tp = arm_compute::NEDepthConcatenateKernel; _Dp = std::default_delete<arm_compute::NEDepthConcatenateKernel []>]'
./arm_compute/runtime/NEON/functions/NEDepthConcatenate.h:44:7:   required from here
/home/user/my-toolchains/aarch64-linux-android-4.9/include/c++/4.9.x/bits/unique_ptr.h:117:22: error: invalid application of 'sizeof' to incomplete type 'arm_compute::NEDepthConcatenateKernel'
  static_assert(sizeof(_Tp)>0,
```
Is there anyone can give me some advices?

**Comment by morgolock** [comment]
Hi @SunAriesCN, could you please share the build command you use?

**Comment by SunAriesCN** [comment]
Hi @morgolock, just as below:
```
aarch64-linux-android-g++ neon_squeezenet/neon_squeezenet.cpp -I. -Iinclude -std=c++11 -L./build -larm_compute-static -L. -o neon_squeezenet.exe ./build/utils/Utils.o -static-libstdc++ -pie
```
In neon_squeezenet.cpp, I just build a squeeznet in it, I think you can recur my problem just with some NEDepthConcatenate declaration. Because I just skip this problem with a header adding to the NEDepthConcatenate.h files. I think there are some flaws inside ComputeLibrary header management. Cloud you mind give me some reason about this issue?

**Comment by morgolock** [comment]
Hi @SunAriesCN,

From the compiler error it seems that you should include arm_compute/core/NEON/NEDepthConcatenateKernel.h in the file neon_squeezenet/neon_squeezenet.cpp.

It seems you include NEFunctions.h in your test, but this header file only brings the declaration of the functions and not the kernels. If you use a kernel in your test you should include it's corresponding header file.

**Comment by SunAriesCN** [comment]
Hi @morgolock, 
No, I think it is not friendly with your developer users, they should not care about this detail. Because it just call something about NEDepthConcatenate class, and it's fail inside the NEDepthConcatenate, users do not always want to care about the implementation of NEDepthConcatenate, let alone the NEDepthConcatenateKernel object. So, I hope the API will be more friendly in the future...

**Comment by morgolock** [comment]
Hi @SunAriesCN,

I'm trying to reproduce this error. We'll fix it.

So you don't use NEDepthConcatenateKernel.h at all in your test? You just call to the NEDephConcatenate function?

**Comment by SunAriesCN** [comment]
Hi @morgolock,
Yes, I just want to call the configure and run functions on my test, and I fix this by using your Kernel header into the other header I mentioned above, but it's something inside ComputeLibrary and I'm not sure if it's always work well.

**Comment by morgolock** [comment]
Hi @SunAriesCN 

Could you please share the source code showing the call to configure?

**Comment by SunAriesCN** [comment]
Hi @SunAriesCN,
Sorry, I can't share my codes to you because of some restriction. But I think you can recur error with some codes as below:
```
#include "arm_compute/runtime/NEON/NEFunctions.h"

#include "arm_compute/core/Types.h"
#include "utils/Utils.h"
#include <iostream>
#include <sstream>
#include <fstream>
#include <ostream>
#include <sys/time.h>
#include <map>

;//some tensor init
;//some data intialization
NEDepthConcatenate fire2_concat;
```

**Comment by morgolock** [comment]
Hi @SunAriesCN 

I managed to reproduce the error, thanks for reporting this. We'll fix it.

**Comment by GeorgeARM** [comment]
Closing as this error should be fixed in 17.09 release. Reopen if needed.
